server/src/main/java/org/candlepin/dto/api/v1/ProductDTO.java

```diff
@@ -1055,6 +1056,15 @@ public int hashCode() {
             }
         }
 
+        int ppHashCode = 0;
+        Collection<ProductDTO> providedProducts = this.getProvidedProducts();
+
+        if (providedProducts != null) {
+            for (ProductDTO dto : providedProducts) {
+                ppHashCode = 31 * ppHashCode + (dto != null ? dto.hashCode() : 0);
+            }
+        }
+
```

Comment:
Similarly to the equals method, where we don't need to check equality in a custom way, we also don't need to calculate hashcode in a custom way. You can just add `this.getProvidedProducts()` in the HashCodeBuilder below.